### ISSUE TITLE
Background Session seems to be broken

### ISSUE THREAD

**Issue by mattisssa** [argmaxinc/WhisperKit#276]
Hey guys, I just tried using the `useBackgroundSession` parameter while downloading a model, and it doesn't seem to be working properly.

Any documentation or details I need to know to configure it? 
Does it work on your side?

**Comment by ZachNagengast** [comment]
Hi @mattisssa, any details on what's not working specifically? Device, OS version, also if possible sample code would be helpful to reproduce

**Comment by mattisssa** [comment]
> Hi @mattisssa, any details on what's not working specifically? Device, OS version, also if possible sample code would be helpful to reproduce

The `useBackgroundSession` generally means that an app can be closed while the model continues to download in the background, correct? This holds true even if the app is later terminated by the system or the user.

I'm testing on an iPhone 15 Pro Max with iOS 18.2. I want to understand what the framework provides under the `useBackgroundSession` parameter. Do I need to handle background tasks manually, or is this implemented in the framework?

**Comment by ZachNagengast** [comment]
This parameter is passed to https://www.github.com/huggingface/swift-transformers here https://github.com/argmaxinc/WhisperKit/blob/2b8737c0cd786714883a47efa617c1d1f364340b/Sources/WhisperKit/Core/WhisperKit.swift#L234 which is then used to download files here https://github.com/huggingface/swift-transformers/blob/37e234e1ee4f2ef9c21c3d5cd592b5ec63d879af/Sources/Hub/Downloader.swift#L38-L42

Curious if you are experiencing a reproducible bug with this?

**Comment by mattisssa** [comment]
> This parameter is passed to https://www.github.com/huggingface/swift-transformers here
> 
> https://github.com/argmaxinc/WhisperKit/blob/2b8737c0cd786714883a47efa617c1d1f364340b/Sources/WhisperKit/Core/WhisperKit.swift#L234
> 
> which is then used to download files here https://github.com/huggingface/swift-transformers/blob/37e234e1ee4f2ef9c21c3d5cd592b5ec63d879af/Sources/Hub/Downloader.swift#L38-L42
> Curious if you are experiencing a reproducible bug with this?

Sure I can walk through the stack and see where it goes but I don't see if whisperkit handles [app suspension](https://developer.apple.com/documentation/foundation/url_loading_system/downloading_files_in_the_background).

---

I'm not sure how the code help you figure out my issue, but setting `useBackgroundSession: true` does not seem to allow downloads to continue in the background when the app goes into the background or is even terminated.

```
extension MLModels {
    actor Loader {
        static var repoName: String { Settings.Defaults.whisperRepoName ?? "argmaxinc/whisperkit-coreml" }

        private let storageURL: URL

        init(storageURL: URL) {
            self.storageURL = storageURL
        }

        func downloadModel(_ model: String, progressCallback: @escaping (DownloadTaskInfo) -> Void) async throws -> URL {
            logger.debug(.whisper, "Model Downloading Started: \(model)")

            let destinationFolder = storageURL.appendingPathComponent(model)

            if FileManager.default.fileExists(atPath: destinationFolder.path) {
                logger.debug(.recognition, "Removing existing model at \(destinationFolder.path)")
                try FileManager.default.removeItem(atPath: destinationFolder.path)
            }

            let startDownloadingDate = Date.now
            return try await withTaskCancellationHandler {
                let downloadedFolder = try await WhisperKit.download(
                    variant: model,
                    from: Self.repoName
                ) { progress in
                    progressCallback(.init(model: model, progress: Float(progress.fractionCompleted), startDownloadingDate: startDownloadingDate))
                }

                try FileManager.default.createDirectory(at: storageURL, withIntermediateDirectories: true )
                try FileManager.default.moveItem(at: downloadedFolder, to: destinationFolder)

                logger.debug(.whisper, "Model Downloaded and Moved to \(destinationFolder.path)")

                return destinationFolder

            } onCancel: { @Sendable in
                logger.debug(.whisper, "Model Downloading Cancelled: \(model)")
            }
        }
    }
}
```

**Comment by ZachNagengast** [comment]
I see, that is helpful. useBackgroundSession is false by default so you'll definitely want it true here:
```swift
               let downloadedFolder = try await WhisperKit.download(
                    variant: model,
                    useBackgroundSession: true,
                    from: Self.repoName
                ) { progress in
                    progressCallback(.init(model: model, progress: Float(progress.fractionCompleted), startDownloadingDate: startDownloadingDate))
                }
```

You can also use this [class method](https://github.com/argmaxinc/WhisperKit/blob/2b8737c0cd786714883a47efa617c1d1f364340b/Sources/WhisperKit/Core/WhisperKit.swift#L281C15-L281C26) on whisperKit to setup the models, as long as `useBackgroundSession` is setup in the config when initializing.


One other thing to check - does your app have background fetch enabled in the capabilities? https://developer.apple.com/documentation/xcode/configuring-background-execution-modes

**Comment by mattisssa** [comment]
I set `useBackgroundSession: true` for sure when I tried to use it 😁

Yes, the app has background fetch enabled and has other background activities and tasks that works fine. 

I've checked the `setupModels` class method but not sure how will it solve my issue. 

For the pure test, I've enabled it for WhisperAX, added background task support, and so on. However, it doesn't work there. 

I opened the app and started downloading a model. After a few minutes, I closed the app and reopened it. This caused the model to be re-downloaded, which I wanted to avoid for the next opening. 

I expected that terminating the app, either by the system or the user, would allow it to download the model in the background. My initial question was whether this is even possible with the framework 🤷‍♂️

**Comment by ZachNagengast** [comment]
Yep it is expected that downloads will continue in the background as long as the app is still alive. Terminating the app by the user though is going to cancel any background tasks, only system terminations allow them to continue in certain situations afaik. You would need to implement some logic to relaunch the app in the background to continue the download, which is not implemented in WhisperKit currently.

**Comment by mattisssa** [comment]
> Yep it is expected that downloads will continue in the background as long as the app is still alive. Terminating the app by the user though is going to cancel any background tasks, only system terminations allow them to continue in certain situations afaik. You would need to implement some logic to relaunch the app in the background to continue the download, which is not implemented in WhisperKit currently.

Ok, thank you for letting me know, that what I needed to clarify 🙌

What is the best way to ask questions regarding the framework? I guess creating Issues is not one of them 😁

**Comment by ZachNagengast** [comment]
@mattisssa We have a discord if you'd like to chat more in depth https://discord.gg/G5F5GZGecC. You can also create a discussion here https://github.com/argmaxinc/WhisperKit/discussions
